sdk/python/kfp/components/modelbase.py

```diff
@@ -0,0 +1,243 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__all__ = [
+    'ModelBase',
```

Comment:
>If I understand correctly, the purpose for this base class is to support class <--> struct <--> serialized form

The "struct <--> serialized form" is handled by a single `yaml.dump` which does it job pretty well.  So the `ModelBase` class only needs to care about the object <-> structure conversions.

>Do we need to support to and from dict? Can it just expose serialize and deserialize to keep the interface simpler?

1. I'm just following the best OSS practices. All the OSS structural serialization/deserialization libraries (swagger, quicktype, jsonpickle) use dicts. I understand the reason why they use dicts and I agree with those reasons. Well, even the DSL compiler converts all objects to dicts and only calls `yaml.dump` in the end.
1. I do not want to bloat the code by inserting `yaml.dump` lines everywhere
1. Most importantly, this won't work in practice. You cannot take a child object serialized as YAML string and insert it as a value for the parent property. Try replacing `from_struct`/`to_struct` with `from_text`/`to_text` on all classes and you'll see it's unworkable.

>Instead of parsing the class members, can we use any existing libs such as jsonpickle?

Two problems with that library:
1. I'm not sure this library has proper support for the schema that's used for components or Kubernetes objects.
2. The main page has big red warning: "**jsonpickle can execute arbitrary Python code. Do not load jsonpickles from untrusted / unauthenticated sources.**"

Have you tried it? You'll see that it does not do what you expect and the purpose of that library is different. What we want is a portable component representation format that can be loaded/saved from python and Go. The purpose of jsonpickle is to dump and load python objects from json text files (as opposed to binary pickle files). While the files generated by jsonpickle are readable, they're very python-bound.

>For example, can each class expose a "validate()" to check fields, and each class override serialize() to first transform the class fields if needed, and then call super's serialize()

That was the old approach. It required lots of code and was error-prone. You had to specify each property in many places:
* Constructor parameter
* Constructor type validation
* self field assignment to function parameter in constructor
* serializer object key, structure field and None check
* deserializer object key, structure field and None check

Instead of big tedious and error-prone boilerplate code, I now have compact data classes that only specify what's needed.
I consider using couple of functions in a base class to be much better than copy-pasting lots of boilerplate code.

If someone develops a library which handles the serialization/deserialization better than the library I wrote, I'll be glad to upgrade. For now, the best tool I considered - quicktype was rather lacking for our goals.